airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/default_file_based_stream.py

```diff
@@ -152,17 +152,18 @@ def _get_raw_json_schema(self) -> JsonSchema:
         elif self.config.schemaless:
             return schemaless_schema
         else:
-            files = self.list_files()
-            total_n_files = len(files)
+            all_files = self.list_files()
+            files_to_read = list(self._cursor.get_files_to_sync(all_files, self.logger))
+            total_n_files = len(files_to_read)
             max_n_files_for_schema_inference = self._discovery_policy.max_n_files_for_schema_inference
             if total_n_files > max_n_files_for_schema_inference:
                 # Use the most recent files for schema inference, so we pick up schema changes during discovery.
-                files = sorted(files, key=lambda x: x.last_modified, reverse=True)[:max_n_files_for_schema_inference]
+                files_to_read = sorted(files_to_read, key=lambda x: x.last_modified, reverse=True)[:max_n_files_for_schema_inference]
                 self.logger.warn(
                     msg=f"Refusing to infer schema for all {total_n_files} files; using {max_n_files_for_schema_inference} files."
                 )
 
-            inferred_schema = self.infer_schema(files)
+            inferred_schema = self.infer_schema(files_to_read)
```

Comment:
If `files_to_read` is empty, mind raising `EMPTY_STREAM` exception before we call `self.infer_schema`? It's a little more specific than the invalid schema exception.